File: reference-implementation/go/session.go
package main

import (
	"context"

	"github.com/google/uuid"
	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

var coreTools = []string{
	"apex.session.*",
	"apex.account.*",
	"apex.order.*",
	"apex.position.*",
	"apex.market.*",
	"apex.risk.*",
}

func registerSessionTools(s *server.MCPServer) {
	registerSessionToolsWithMode(s, state, false)
}

func registerSessionToolsWithMode(s *server.MCPServer, st *referenceState, httpMode bool) {
	s.AddTool(
		mcp.NewTool("apex.session.authenticate",
			mcp.WithDescription("Establish an authenticated trading session. The broker validates credentials directly and binds the result to the MCP session."),
			mcp.WithString("token", mcp.Required(), mcp.Description("Broker-issued JWT or OAuth token")),
			mcp.WithString("token_type", mcp.Description("Token type: jwt or oauth2"), mcp.Enum("jwt", "oauth2")),
			mcp.WithString("account_id", mcp.Description("Optional — broker may derive from token")),
			mcp.WithString("hub_session_id", mcp.Description("Optional session reference from caller")),
			mcp.WithReadOnlyHintAnnotation(false),
			mcp.WithDestructiveHintAnnotation(false),
			mcp.WithIdempotentHintAnnotation(true),
		),
		func(_ context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			args := request.GetArguments()
			token := strParam(args, "token", "")
			if len(token) < 10 {
				return jsonResult(apexError("APEX_4001", "auth", "Invalid or expired token"))
			}

			// Trigger onAuthenticated callback (starts tick engine in HTTP mode)
			if st.onAuthenticated != nil {
				st.onAuthenticated()
			}

			return jsonResult(sessionResponse{
				SessionID:    uuid.NewString(),
				AccountID:    strParam(args, "account_id", "ACC_12345"),
				ExpiresAt:    hoursFromNow(1),
				Capabilities: coreTools,
				Profiles:     []string{"fx"},
				BrokerID:     "reference-broker",
				BrokerName:   "APEX Reference Broker",
			})
		},
	)

	s.AddTool(
		mcp.NewTool("apex.session.capabilities",
			mcp.WithDescription("Query the full capability manifest of this broker implementation."),
			mcp.WithReadOnlyHintAnnotation(true),
			mcp.WithDestructiveHintAnnotation(false),
			mcp.WithIdempotentHintAnnotation(true),
		),
		func(_ context.Context, _ mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			var realtimeContract map[string]any
			if httpMode {
				realtimeContract = map[string]any{
					"transport_mode":         "streamable_http",
					"reconnect_mode":         "session_replay",
					"max_retention_events":   10000,
					"max_retention_seconds":  0,
					"quote_freshness_ms":     1000,
					"account_freshness_ms":   2000,
					"tick_interval_ms":       2000,
					"notifications": []string{
						"notifications/apex.order.filled",
						"notifications/apex.order.partially_filled",
						"notifications/apex.order.rejected",
						"notifications/apex.market.candle_closed",
						"notifications/apex.risk.kill_switch_engaged",
						"notifications/apex.session.replay_failed",
						"notifications/apex.session.gap_fill",
					},
				}
			} else {
				realtimeContract = map[string]any{
					"transport_mode":       "stdio",
					"reconnect_mode":       "no_replay",
					"quote_freshness_ms":   1000,
					"account_freshness_ms": 2000,
				}
			}

			return jsonResult(capabilitiesResponse{
				ApexVersion:         serverVersion,
				BrokerID:            "reference-broker",
				CoreTools:           coreTools,
				Profiles:            map[string]string{"fx": serverVersion},
				VendorExtensions:    nil,
				RateLimits:          map[string]int{"orders_per_second": 10, "market_data_per_second": 100},
				SupportedOrderTypes:  []string{"market", "limit", "stop", "stop_limit"},
				SupportedTif:         []string{"GTC", "IOC", "FOK", "DAY"},
				ProductionProfiles:   map[string]bool{"realtime": true, "autonomous": false},
				RealtimeContract:     realtimeContract,
			})
		},
	)

	s.AddTool(
		mcp.NewTool("apex.session.heartbeat",
			mcp.WithDescription("Keep-alive ping. Hub marks session degraded if response exceeds 500ms."),
			mcp.WithString("timestamp", mcp.Required(), mcp.Description("ISO8601 timestamp")),
			mcp.WithReadOnlyHintAnnotation(true),
			mcp.WithDestructiveHintAnnotation(false),
			mcp.WithIdempotentHintAnnotation(true),
		),
		handleSessionHeartbeat,
	)

	s.AddTool(
		mcp.NewTool("apex.session.acknowledge",
			mcp.WithDescription("Acknowledge receipt of events through last_event_id, allowing the broker to trim the replay buffer."),
			mcp.WithString("last_event_id", mcp.Required(), mcp.Description("The ID of the last event the client has successfully processed")),
			mcp.WithReadOnlyHintAnnotation(true),
			mcp.WithDestructiveHintAnnotation(false),
			mcp.WithIdempotentHintAnnotation(true),
		),
		func(_ context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			args := request.GetArguments()
			lastEventID := strParam(args, "last_event_id", "")
			if lastEventID == "" {
				return jsonResult(apexError("APEX_4011", "validation", "last_event_id is required"))
			}
			if st.replayBuffer == nil {
				// No-op in stdio mode (no replay buffer)
				return jsonResult(map[string]any{
					"acknowledged_through": "0",
					"buffer_depth":         0,
				})
			}
			ackThrough, depth := st.replayBuffer.Acknowledge(lastEventID)
			return jsonResult(map[string]any{
				"acknowledged_through": ackThrough,
				"buffer_depth":         depth,
			})
		},
	)

	s.AddTool(
		mcp.NewTool("reference.test.set_realtime_state",
			mcp.WithDescription("Reference-only fault injection for conformance and resilience testing."),
			mcp.WithBoolean("quote_stale"),
			mcp.WithBoolean("risk_stale"),
			mcp.WithBoolean("force_sequence_gap"),
			mcp.WithBoolean("kill_switch_active"),
			mcp.WithBoolean("partial_fill_next_order"),
		),
		func(_ context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			args := request.GetArguments()
			quoteStale := boolPointer(args, "quote_stale")
			riskStale := boolPointer(args, "risk_stale")
			forceSequenceGap := boolPointer(args, "force_sequence_gap")
			killSwitchActive := boolPointer(args, "kill_switch_active")
			partialFillNext := boolPointer(args, "partial_fill_next_order")

			return jsonResult(map[string]any{
				"ok":     true,
				"faults": st.setRealtimeFaults(quoteStale, riskStale, forceSequenceGap, killSwitchActive, partialFillNext),
			})
		},
	)
}

func handleSessionHeartbeat(_ context.Context, _ mcp.CallToolRequest) (*mcp.CallToolResult, error) {
	return jsonResult(struct {
		Timestamp string `json:"timestamp"`
		Status    string `json:"status"`
	}{
		Timestamp: nowISO(),
		Status:    "ok",
	})
}

File: reference-implementation/typescript/src/tools/session.ts
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";

import {
  CORE_TOOL_CAPABILITIES,
  SERVER_VERSION,
  SUPPORTED_ORDER_TYPES,
  SUPPORTED_PROFILES,
  SUPPORTED_TIF,
} from "../lib/constants.js";
import { apexError, hoursFromNow, nowIso } from "../lib/helpers.js";
import type { ReferenceTradingState } from "../lib/resources.js";
import type { ReplayBuffer } from "../lib/replay-buffer.js";
import { z } from "zod";

export interface SessionToolOptions {
  transportMode?: "streamable_http";
  onAuthenticated?: () => void;
  replayBuffer?: ReplayBuffer;
}

export function registerSessionTools(server: McpServer, state: ReferenceTradingState, options?: SessionToolOptions): void {
  const transportMode = options?.transportMode ?? "streamable_http";
  server.registerTool(
    "apex.session.authenticate",
    {
      description:
        "Establish an authenticated trading session. The broker validates credentials directly and binds the result to the MCP session.",
      inputSchema: {
        token: z.string().describe("Broker-issued JWT or OAuth token"),
        token_type: z.enum(["jwt", "oauth2"]).default("jwt"),
        account_id: z.string().optional().describe("Optional — broker may derive from token"),
        hub_session_id: z.string().optional().describe("Optional session reference from caller"),
      },
      annotations: { readOnlyHint: false, destructiveHint: false, idempotentHint: true },
    },
    async ({ token, account_id }) => {
      if (token.length < 10) {
        return {
          structuredContent: apexError("APEX_4001", "auth", "Invalid or expired token"),
          content: [],
        };
      }

      options?.onAuthenticated?.();

      return {
        structuredContent: {
          session_id: crypto.randomUUID(),
          account_id: account_id ?? "ACC_12345",
          expires_at: hoursFromNow(1),
          capabilities: [...CORE_TOOL_CAPABILITIES],
          profiles: [...SUPPORTED_PROFILES],
          broker_id: "reference-broker",
          broker_name: "APEX Reference Broker",
        },
        content: [],
      };
    },
  );

  server.registerTool(
    "apex.session.capabilities",
    {
      description: "Query the full capability manifest of this broker implementation.",
      inputSchema: {},
      annotations: { readOnlyHint: true, destructiveHint: false, idempotentHint: true },
    },
    async () => ({
      structuredContent: {
        apex_version: SERVER_VERSION,
        broker_id: "reference-broker",
        core_tools: [...CORE_TOOL_CAPABILITIES],
        profiles: { fx: SERVER_VERSION },
        vendor_extensions: null,
        rate_limits: { orders_per_second: 10, market_data_per_second: 100 },
        supported_order_types: [...SUPPORTED_ORDER_TYPES],
        supported_tif: [...SUPPORTED_TIF],
        production_profiles: {
          realtime: true,
          autonomous: false,
        },
        realtime_contract: {
          transport_mode: transportMode,
          reconnect_mode: "session_replay",
          max_retention_events: 10000,
          max_retention_seconds: 0,
          quote_freshness_ms: 1000,
          account_freshness_ms: 2000,
          tick_interval_ms: 2000,
          notifications: [
            "notifications/apex.order.filled",
            "notifications/apex.order.partially_filled",
            "notifications/apex.order.rejected",
            "notifications/apex.market.candle_closed",
            "notifications/apex.risk.kill_switch_engaged",
            "notifications/apex.session.replay_failed",
            "notifications/apex.session.gap_fill",
          ],
        },
      },
      content: [],
    }),
  );

  server.registerTool(
    "apex.session.heartbeat",
    {
      description: "Keep-alive ping. Hub marks session degraded if response exceeds 500ms.",
      inputSchema: { timestamp: z.string().describe("ISO8601 timestamp") },
      annotations: { readOnlyHint: true, destructiveHint: false, idempotentHint: true },
    },
    async () => ({
      structuredContent: { timestamp: nowIso(), status: "ok" },
      content: [],
    }),
  );

  server.registerTool(
    "apex.session.acknowledge",
    {
      description: "Acknowledge receipt of SSE events. Server discards acknowledged events.",
      inputSchema: {
        last_event_id: z.string().describe("Last SSE event ID processed"),
      },
      annotations: { readOnlyHint: true, destructiveHint: false, idempotentHint: true },
    },
    async ({ last_event_id }) => {
      if (!last_event_id) {
        return {
          structuredContent: apexError("APEX_4011", "validation", "last_event_id is required"),
          content: [],
        };
      }
      const result = options?.replayBuffer
        ? options.replayBuffer.acknowledge(last_event_id)
        : { acknowledged_through: "0", buffer_depth: 0 };
      return {
        structuredContent: result,
        content: [],
      };
    },
  );

  server.registerTool(
    "reference.test.set_realtime_state",
    {
      description: "Reference-only fault injection for conformance and resilience testing.",
      inputSchema: {
        quote_stale: z.boolean().optional(),
        risk_stale: z.boolean().optional(),
        force_sequence_gap: z.boolean().optional(),
        kill_switch_active: z.boolean().optional(),
        partial_fill_next_order: z.boolean().optional(),
      },
      annotations: { readOnlyHint: false, destructiveHint: false, idempotentHint: true },
    },
    async (input) => {
      state.setRealtimeFaults(input);
      return {
        structuredContent: {
          ok: true,
          faults: state.currentFaults(),
        },
        content: [],
      };
    },
  );
}
